Add guide for "in parallel"

rakina opened the pull request:
Adds a new "Be aware that navigations can happen between steps running in parallel" section.

PTAL @annevk! I'm not really sure what I'm saying is quite right, so please correct me if necessary :)

rakina committed:
Add guide for "in parallel"

Adds a new "Be aware that navigations can happen between steps running in parallel" section.

Fixes #2.

rakina committed:
Add guide for "in parallel"

Adds a new "Be aware that navigations can happen between steps running in parallel" section.

Fixes #2.

annevk commented on index.bs:
```diff
@@ -221,6 +221,23 @@ and doing cleanup as necessary (see the [[#listen-fully-active]] pattern above).
 
+<h4 id="non-synchronous-steps">Be aware that navigations can happen between steps running [=in parallel=]</h4>
+While running steps [=in parallel=],
+the document the steps are running on or are otherwise using might be navigated away and navigated to again.
```
```suggestion
the document the steps are running in or are otherwise using might be navigated away and (later) navigated to again.
```

annevk commented on index.bs:
```diff
@@ -221,6 +221,23 @@ and doing cleanup as necessary (see the [[#listen-fully-active]] pattern above).
+<h4 id="non-synchronous-steps">Be aware that navigations can happen between steps running [=in parallel=]</h4>
+While running steps [=in parallel=],
+the document the steps are running on or are otherwise using might be navigated away and navigated to again.
+The steps themselves will not run while the document is not [=fully active=],
```
Actually, they will. There's nothing stopping in parallel steps at the moment. What won't run is steps in a queued task if that task is associated with the document.

rakina replied:
Ah, right. I've rewritten this to mention that possibility, and add a suggestion to maybe not run some steps while the document is fully active. PTAL again!

annevk reviewed (commented):
Thanks for tackling this @rakina!

cc @domenic

rakina committed:
Apply suggestion

Co-authored-by: Anne van Kesteren <annevk@annevk.nl>

rakina committed:
Apply suggestion

Co-authored-by: Anne van Kesteren <annevk@annevk.nl>

rakina committed:
Steps might run when document is not fully active

rakina committed:
Steps might run when document is not fully active

annevk commented on index.bs:
```diff
@@ -221,6 +221,25 @@ and doing cleanup as necessary (see the [[#listen-fully-active]] pattern above).
 </div>
 
+<h4 id="non-synchronous-steps">Be aware that navigations can happen between steps running [=in parallel=]</h4>
+While running steps [=in parallel=],
+the document the steps are running in or are otherwise using might be navigated away and (later) navigated to again.
```
This is still confusing. The document represents the conceptual "main thread". "In parallel" is a concept representing another thread/process. So if the steps run in parallel, they are by definition not running on the main thread.

Task queued steps might, of course, but then you need to spell that out. As you do below.

rakina replied:
OK, how about:
"While running steps [=in parallel=], if the steps are in a queued task associated with the document,
the document might be navigated away and (later) on navigated to again. The steps themselves might not run while the document is not [=fully active=], and a considerable amount of time might have passed between the steps,
so the user might not expect some states that are saved in the steps to be preserved. "

annevk replied:
Still unclear to me. Is this what you mean:

> If steps that run in parallel queue a task that is associated with a document, you need to be aware that the document might have been navigated away from and (later on) navigated to again. This can happen while running the steps in parallel and can have happened when the steps in the queued task eventually run. ...

?

rakina replied:
Thanks! I used that now, lmk if there are things that still need editing :)

annevk commented on index.bs:
```diff
@@ -221,6 +221,25 @@ and doing cleanup as necessary (see the [[#listen-fully-active]] pattern above).
+  APIs that request permission to the user, or requests user interaction through prompt, etc.,
+  should consider re-requesting the user if it detects that the document had become non-fully active.
+</div>
+Note: In theory, steps that run [=in parallel=] should generally already be resilient to changes that might occur between its steps.
```
"should" in a note is generally not allowed, but maybe conformance requirements don't apply here? https://infra.spec.whatwg.org/#conformance

rakina replied:
Ah OK, maybe "are probably already resilient" will be better?

annevk replied:
"ought to be"

rakina replied:
Used that now, thanks!

rakina committed:
Use annevk's suggestions

rakina committed:
Use annevk's suggestions

annevk reviewed (approved):
Thanks, this looks good to me. If @domenic has a minute maybe he can do a quick skim as well as I tend to get review blind after several iterations. 😅

domenic commented on index.bs:
```diff
@@ -221,6 +221,24 @@ and doing cleanup as necessary (see the [[#listen-fully-active]] pattern above).
+  APIs that request permission to the user, or requests user interaction through prompt, etc.,
+  should consider re-requesting the user if it detects that the document had become non-fully active.
+</div>
+Note: In theory, steps that run [=in parallel=] ought to be already resilient to changes that might occur between its steps.
```
Personally I would move this note up above the "When writing steps..." pattern.

rakina replied:
Moved now!

domenic commented on .DS_Store:
I suggest adding `.DS_Store` to your global gitignore, and not committing this :)

domenic commented on index.bs:
```diff
@@ -221,6 +221,24 @@ and doing cleanup as necessary (see the [[#listen-fully-active]] pattern above).
+<div class=example>
+  APIs that request permission to the user, or requests user interaction through prompt, etc.,
+  should consider re-requesting the user if it detects that the document had become non-fully active.
+</div>
```
I wish we had a better example here. I don't think this is particularly actionable for spec authors, e.g., I don't see how the advice here would change how https://w3c.github.io/permissions/#dfn-request-permission-to-use is written.

Note that because that algorithm queues a task, and tasks don't run on non-fully active documents, the result would be that the permissions store is updated after being reactivated from bfcache. That seems OK?

@annevk, did you have other examples of problematic cases you've seen in various specs?

annevk replied:
Well, there's no guarantee the document is still fully active when "ask the user" happens, right? I think that's the potentially more problematic aspect.

rakina replied:
An example @annevk mentioned in the old PR was https://storage.spec.whatwg.org/#dom-storagemanager-persist, the steps that run in the "in parallel" block there can have navigations/bfcache restores in between them. Maybe we should mention that specific example?

domenic replied:
I guess I'm just still unsure on what the concrete guidance is. Since @annevk is the editor of the Storage Standard, does he have suggestions on how that algorithm should be updated to handle this concern?

annevk replied:
I don't, but ...

Perhaps asking the user should itself be a queued task of sorts or at least something that is able to check the document is still current and the user can answer the question. And then you either resume from there or you don't.

And if you resume you get to update state and queue another task. Which then potentially has to deal with further navigation thingies.

@johannhof might have thoughts on this having done some recent work on Permissions.

domenic replied:
Yeah, so, I feel like if neither you nor I can figure out how to apply the advice here to any existing specs, this advice isn't going to be that useful for other spec authors... So I wonder why we're adding it.

annevk replied:
That is one way of approaching it. Another way of approaching it is whether we agree there is a problem to be solved here.

rakina committed:
gitignore

rakina committed:
gitignore

rakina committed:
move note up

rakina committed:
move note up

domenic commented on .gitignore:
I meant your system-global gitignore, not this repository specifically :)

rakina committed:
remove gitignore

rakina committed:
remove gitignore

domenic reviewed (approved):
LGTM, although still curious about whether we could have better examples. Might be best to merge and do that as a followup though...

rakina committed:
Add references to steps in HTML spec for fully active state changes

This makes the guide more clear on what to do exactly to listen to changes to the fully active state change.

rakina committed:
Use exported "reactivate a Document"

rakina committed:
Add references to steps in HTML spec to listen to fully active state changes

hober commented:
@rakina, can you resolve the conflicts on this PR? I think it's ready to land otherwise.

rakina committed:
resolve conflict

rakina committed:
Merge branch 'branch-to-review' of https://github.com/w3ctag/bfcache-guide into branch-to-review

rakina commented:
@hober done, although I'm not sure if the discussion between @domenic and @annevk at https://github.com/w3ctag/bfcache-guide/pull/3/files#r1148723589 is resolved yet.

jyasskin commented:
Sorry for taking a long time to get back to this. Several of us discussed this in our breakout today, and we're wondering if this whole guide ought to move to somewhere in the HTML spec. @martinthomson and I felt that the guide is hard to discover in its current location, even though https://w3ctag.github.io/design-principles/#support-non-fully-active mentions it, and that HTML should really introduce the concept. https://html.spec.whatwg.org/multipage/document-sequences.html#fully-active might be a good section, but the HTML editors might also have a better idea. @annevk @domenic, would y'all accept patches to move parts or all of this guide into HTML?

We're also happy to merge this PR into its current location if and when @domenic and @annevk agree on its contents, but we hope that can be a stepping stone to putting it closer to the normative specification.

I also want to acknowledge that @rakina has done some great work on this guide, and that it's not fair to ask her to pick it up again after a 2-year delay, so even if there's agreement in theory to move the guide, we won't be surprised or disappointed if we need to find a new person to do the actual work.
